File: apps/yryvu-app/src/ipc/rebase.ts
// SPDX-License-Identifier: AGPL-3.0-or-later

import { invoke } from "@tauri-apps/api/core";

export type RebaseActionKind = "pick" | "reword" | "edit" | "squash" | "fixup" | "drop";

export type PauseReason = "edit" | "conflict";

export interface CommitSummary {
  oid: string;
  short_oid: string;
  summary: string;
  author_name: string;
  author_email: string;
}

export interface RebaseStep {
  oid: string;
  action: RebaseActionKind;
  new_message?: string | null;
}

export interface RebasePlan {
  onto: string;
  steps: RebaseStep[];
}

export interface RebaseState {
  onto: string;
  steps: RebaseStep[];
  current_step: number;
  original_head: string;
  head_branch: string | null;
  pause_reason: PauseReason | null;
}

export function listCommitsForRebase(repoPath: string, upstream: string) {
  return invoke<CommitSummary[]>("list_commits_for_rebase", { repoPath, upstream });
}

export function beginInteractiveRebase(repoPath: string, plan: RebasePlan) {
  return invoke<RebaseState>("begin_interactive_rebase", { repoPath, plan });
}

export function continueInteractiveRebase(repoPath: string) {
  return invoke<RebaseState>("continue_interactive_rebase", { repoPath });
}

export function skipInteractiveRebaseStep(repoPath: string) {
  return invoke<RebaseState>("skip_interactive_rebase_step", { repoPath });
}

export function abortInteractiveRebase(repoPath: string) {
  return invoke<void>("abort_interactive_rebase", { repoPath });
}

export function getInteractiveRebaseState(repoPath: string) {
  return invoke<RebaseState | null>("get_interactive_rebase_state", { repoPath });
}

File: crates/yryvu-bridge/src/commands/rebase.rs
// SPDX-License-Identifier: AGPL-3.0-or-later

use std::path::PathBuf;

use crate::backend::GitBackend;
use crate::repo::rebase::interactive::{CommitSummary, RebasePlan, RebaseState};
use crate::repo::GixBackend;

#[tauri::command]
pub async fn list_commits_for_rebase(
    repo_path: String,
    upstream: String,
) -> Result<Vec<CommitSummary>, String> {
    tauri::async_runtime::spawn_blocking(move || {
        GixBackend
            .list_commits_for_rebase(&PathBuf::from(&repo_path), &upstream)
            .map_err(|e| e.to_string())
    })
    .await
    .map_err(|e| e.to_string())?
}

#[tauri::command]
pub async fn begin_interactive_rebase(
    repo_path: String,
    plan: RebasePlan,
) -> Result<RebaseState, String> {
    tauri::async_runtime::spawn_blocking(move || {
        GixBackend
            .begin_interactive_rebase(&PathBuf::from(&repo_path), plan)
            .map_err(|e| e.to_string())
    })
    .await
    .map_err(|e| e.to_string())?
}

#[tauri::command]
pub async fn continue_interactive_rebase(repo_path: String) -> Result<RebaseState, String> {
    tauri::async_runtime::spawn_blocking(move || {
        GixBackend
            .continue_interactive_rebase(&PathBuf::from(&repo_path))
            .map_err(|e| e.to_string())
    })
    .await
    .map_err(|e| e.to_string())?
}

#[tauri::command]
pub async fn skip_interactive_rebase_step(repo_path: String) -> Result<RebaseState, String> {
    tauri::async_runtime::spawn_blocking(move || {
        GixBackend
            .skip_interactive_rebase_step(&PathBuf::from(&repo_path))
            .map_err(|e| e.to_string())
    })
    .await
    .map_err(|e| e.to_string())?
}

#[tauri::command]
pub async fn abort_interactive_rebase(repo_path: String) -> Result<(), String> {
    tauri::async_runtime::spawn_blocking(move || {
        GixBackend
            .abort_interactive_rebase(&PathBuf::from(&repo_path))
            .map_err(|e| e.to_string())
    })
    .await
    .map_err(|e| e.to_string())?
}

#[tauri::command]
pub async fn get_interactive_rebase_state(
    repo_path: String,
) -> Result<Option<RebaseState>, String> {
    tauri::async_runtime::spawn_blocking(move || {
        GixBackend
            .get_interactive_rebase_state(&PathBuf::from(&repo_path))
            .map_err(|e| e.to_string())
    })
    .await
    .map_err(|e| e.to_string())?
}
